src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -5788,73 +5968,113 @@ private bool CheckInterpolatedStringHandlerConversionEscape(BoundExpression expr
 #if DEBUG
             _visited = previousVisited;
 #endif
 
-            var arguments = ArrayBuilder<BoundExpression>.GetInstance();
-            GetInterpolatedStringHandlerArgumentsForEscape(expression, arguments);
+            // Narrow the scope for implicit calls which allow the receiver to capture refs from the arguments.
+            result = result && CheckValEscapeOfInterpolatedStringHandlerCalls(expression, escapeFrom, escapeTo, diagnostics);
 
-            bool result = true;
-            foreach (var argument in arguments)
+            return result;
+        }
+
+        private SafeContext GetValEscapeOfInterpolatedStringHandlerCalls(BoundExpression expression, SafeContext localScopeDepth)
+        {
+            SafeContext scope = SafeContext.CallingMethod;
+
+            while (true)
             {
-                if (!CheckValEscape(argument.Syntax, argument, escapeFrom, escapeTo, checkingReceiver: false, diagnostics))
+                switch (expression)
                 {
-                    result = false;
-                    break;
+                    case BoundBinaryOperator binary:
+                        scope = scope.Intersect(GetValEscapeOfInterpolatedStringHandlerCalls(binary.Right, localScopeDepth));
+                        expression = binary.Left;
+                        break;
+
+                    case BoundInterpolatedString interpolatedString:
+                        return scope.Intersect(getParts(interpolatedString, localScopeDepth));
+
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(expression.Kind);
                 }
             }
 
-            arguments.Free();
-            return result;
+            SafeContext getParts(BoundInterpolatedString interpolatedString, SafeContext localScopeDepth)
+            {
+                SafeContext scope = SafeContext.CallingMethod;
+
+                foreach (var part in interpolatedString.Parts)
+                {
+                    if (part is not BoundCall call)
+                    {
+                        // Dynamic calls cannot have ref struct parameters.
+                        continue;
+                    }
+
+                    scope = scope.Intersect(GetInvocationEscapeToReceiver(
+                        MethodInfo.Create(call.Method),
+                        call.Method.Parameters,
+                        call.Arguments,
+                        call.ArgumentRefKindsOpt,
+                        call.ArgsToParamsOpt,
+                        localScopeDepth));
+                }
+
+                return scope;
+            }
         }
 
-        private void GetInterpolatedStringHandlerArgumentsForEscape(BoundExpression expression, ArrayBuilder<BoundExpression> arguments)
+        private bool CheckValEscapeOfInterpolatedStringHandlerCalls(BoundExpression expression, SafeContext escapeFrom, SafeContext escapeTo, BindingDiagnosticBag diagnostics)
```

Comment:
Does this need to take in a `BoundExpression`? Or would it work to take in a `BoundInterpolatedString` and make the `expression` used for traversal a local?

Reply:
The expression can _start_ as `BoundBinaryOperator` or `BoundInterpolatedString`. So yes, this needs to take `BoundExpression`.